Automated TemplateCombination's methods GET and DELETE

renzon opened the pull request:
Test results on sat 6.3:

```console
============================= test session starts ==============================
platform linux2 -- Python 2.7.12, pytest-3.0.3, py-1.4.31, pluggy-0.4.0
rootdir: /home/renzo/PycharmProjects/robottelo, inifile: 
plugins: xdist-1.15.0, cov-2.3.1
collected 1 items

test_template_combination.py 2016-12-16 09:18:41 - robottelo - DEBUG - Started setUpClass: tests.foreman.api.test_template_combination/TestTemplateCombination
2016-12-16 09:18:41 - nailgun.client - DEBUG - Making HTTP POST request to https://foo.bar/api/v2/hostgroups with options {'verify': False, 'auth': ('baz', '*****'), 'headers': {'content-type': 'application/json'}} and data {"hostgroup": {"name": "YtYTfdODe"}}.
2016-12-16 09:18:42 - nailgun.client - DEBUG - Received HTTP 201 response: {"subnet_id":null,"subnet_name":null,"operatingsystem_id":null,"operatingsystem_name":null,"domain_id":null,"domain_name":null,"environment_id":null,"environment_name":null,"compute_profile_id":null,"compute_profile_name":null,"ancestry":null,"parent_id":null,"parent_name":null,"puppet_proxy_id":null,"puppet_proxy_name":null,"puppet_ca_proxy_id":null,"puppet_ca_proxy_name":null,"ptable_id":null,"ptable_name":null,"medium_id":null,"medium_name":null,"architecture_id":null,"architecture_name":null,"realm_id":null,"realm_name":null,"created_at":"2016-12-16 11:18:42 UTC","updated_at":"2016-12-16 11:18:42 UTC","id":4,"name":"YtYTfdODe","title":"YtYTfdODe","template_combinations":[],"puppetclasses":[],"config_groups":[],"parameters":[],"all_puppetclasses":[],"locations":[],"organizations":[]}
2016-12-16 09:18:42 - nailgun.client - DEBUG - Making HTTP GET request to https://foo.bar/api/v2/hostgroups/4 with options {'verify': False, 'auth': ('baz', '*****'), 'headers': {'content-type': 'application/json'}} and no data.
2016-12-16 09:18:43 - nailgun.client - DEBUG - Received HTTP 200 response: {"content_source_id":null,"content_source_name":null,"content_view_id":null,"content_view_name":null,"lifecycle_environment_id":null,"lifecycle_environment_name":null,"subnet_id":null,"subnet_name":null,"operatingsystem_id":null,"operatingsystem_name":null,"domain_id":null,"domain_name":null,"environment_id":null,"environment_name":null,"compute_profile_id":null,"compute_profile_name":null,"ancestry":null,"parent_id":null,"parent_name":null,"puppet_proxy_id":null,"puppet_proxy_name":null,"puppet_ca_proxy_id":null,"puppet_ca_proxy_name":null,"ptable_id":null,"ptable_name":null,"medium_id":null,"medium_name":null,"architecture_id":null,"architecture_name":null,"realm_id":null,"realm_name":null,"created_at":"2016-12-16 11:18:42 UTC","updated_at":"2016-12-16 11:18:42 UTC","id":4,"name":"YtYTfdODe","title":"YtYTfdODe","template_combinations":[],"puppetclasses":[],"config_groups":[],"parameters":[],"all_puppetclasses":[],"locations":[],"organizations":[]}
2016-12-16 09:18:43 - nailgun.client - DEBUG - Making HTTP POST request to https://foo.bar/api/v2/environments with options {'verify': False, 'auth': ('baz', '*****'), 'headers': {'content-type': 'application/json'}} and data {"environment": {"name": "9pmV69"}}.
2016-12-16 09:18:44 - nailgun.client - DEBUG - Received HTTP 201 response: {"created_at":"2016-12-16 11:18:44 UTC","updated_at":"2016-12-16 11:18:44 UTC","name":"9pmV69","id":5,"template_combinations":[],"puppetclasses":[],"locations":[],"organizations":[]}
2016-12-16 09:18:44 - robottelo - DEBUG - Started Test: TestTemplateCombination/test_positive_get_combination
2016-12-16 09:18:44 - nailgun.client - DEBUG - Making HTTP POST request to https://foo.bar/api/v2/config_templates with options {'verify': False, 'auth': ('baz', '*****'), 'headers': {'content-type': 'application/json'}} and data {"config_template": {"template_combinations_attributes": [{"environment_id": 5, "hostgroup_id": 4}], "snippet": false, "template_kind_id": 1, "template": "\u6d72\ud867\udff0", "name": "WNqaYlQshse"}}.
2016-12-16 09:18:45 - nailgun.client - DEBUG - Received HTTP 201 response: {"template":"浲𩿰","locked":false,"snippet":false,"audit_comment":null,"created_at":"2016-12-16 11:18:45 UTC","updated_at":"2016-12-16 11:18:45 UTC","id":109,"name":"WNqaYlQshse","template_kind_id":1,"template_kind_name":"Bootdisk","template_combinations":[{"id":4,"provisioning_template_id":109,"provisioning_template_name":"WNqaYlQshse","hostgroup_id":4,"hostgroup_name":"YtYTfdODe","environment_id":5,"environment_name":"9pmV69","config_template_id":109,"config_template_name":"WNqaYlQshse"}],"operatingsystems":[],"os_default_templates":[],"locations":[],"organizations":[]}
2016-12-16 09:18:45 - robottelo.decorators - INFO - Bugzilla bug 1369737 not in cache. Fetching.
2016-12-16 09:18:50 - robottelo.ssh - INFO - Instantiated Paramiko client 0x7fe0fceb5450
2016-12-16 09:18:50 - robottelo.ssh - DEBUG - Connected to [foo.bar]
2016-12-16 09:18:50 - robottelo.ssh - DEBUG - >>> rpm -q satellite
2016-12-16 09:18:51 - robottelo.ssh - DEBUG - <<< stdout
satellite-6.3.0-11.1.beta.el7sat.noarch

2016-12-16 09:18:51 - robottelo.ssh - INFO - Destroying Paramiko client 0x7fe0fceb5450
2016-12-16 09:18:51 - robottelo.ssh - INFO - Destroyed Paramiko client 0x7fe0fceb5450
2016-12-16 09:18:51 - robottelo.host_info - DEBUG - Host Satellite version: 6.3
2016-12-16 09:18:51 - nailgun.client - DEBUG - Making HTTP GET request to https://foo.bar/api/v2/template_combinations/4 with options {'verify': False, 'auth': ('baz', '*****'), 'headers': {'content-type': 'application/json'}} and no data.
2016-12-16 09:18:52 - nailgun.client - DEBUG - Received HTTP 200 response: {"created_at":"2016-12-16 11:18:45 UTC","updated_at":"2016-12-16 11:18:45 UTC","id":4,"provisioning_template_id":109,"provisioning_template_name":"WNqaYlQshse","hostgroup_id":4,"hostgroup_name":"YtYTfdODe","environment_id":5,"environment_name":"9pmV69","config_template_id":109,"config_template_name":"WNqaYlQshse"}
2016-12-16 09:18:52 - robottelo - DEBUG - Finished Test: TestTemplateCombination/test_positive_get_combination
2016-12-16 09:18:52 - nailgun.client - DEBUG - Making HTTP DELETE request to https://foo.bar/api/v2/template_combinations/4 with options {'verify': False, 'auth': ('baz', '*****'), 'headers': {'content-type': 'application/json'}} and no data.
2016-12-16 09:18:53 - nailgun.client - DEBUG - Received HTTP 200 response: {"id":4,"provisioning_template_id":109,"hostgroup_id":4,"environment_id":5,"created_at":"2016-12-16T11:18:45.688Z","updated_at":"2016-12-16T11:18:45.688Z"}
2016-12-16 09:18:53 - nailgun.client - DEBUG - Making HTTP DELETE request to https://foo.bar/api/v2/config_templates/109 with options {'verify': False, 'auth': ('baz', '*****'), 'headers': {'content-type': 'application/json'}} and no data.
2016-12-16 09:18:54 - nailgun.client - DEBUG - Received HTTP 200 response: {"id":109,"name":"WNqaYlQshse","template":"浲𩿰","snippet":false,"template_kind_id":1,"created_at":"2016-12-16T11:18:45.653Z","updated_at":"2016-12-16T11:18:45.653Z","locked":false,"default":false,"vendor":null,"os_family":null,"job_category":"Miscellaneous","provider_type":null,"description_format":null}
.2016-12-16 09:18:54 - robottelo - DEBUG - Started tearDownClass: tests.foreman.api.test_template_combination/TestTemplateCombination
2016-12-16 09:18:54 - nailgun.client - DEBUG - Making HTTP DELETE request to https://foo.bar/api/v2/hostgroups/4 with options {'verify': False, 'auth': ('baz', '*****'), 'headers': {'content-type': 'application/json'}} and no data.
2016-12-16 09:18:55 - nailgun.client - DEBUG - Received HTTP 200 response: {"id":4,"name":"YtYTfdODe","created_at":"2016-12-16T11:18:42.184Z","updated_at":"2016-12-16T11:18:42.184Z","environment_id":null,"operatingsystem_id":null,"architecture_id":null,"medium_id":null,"ptable_id":null,"root_pass":null,"puppet_ca_proxy_id":null,"use_image":null,"image_file":"","ancestry":null,"vm_defaults":null,"subnet_id":null,"domain_id":null,"puppet_proxy_id":null,"title":"YtYTfdODe","realm_id":null,"compute_profile_id":null,"content_source_id":null,"grub_pass":"","content_view_id":null,"lifecycle_environment_id":null,"lookup_value_matcher":"hostgroup=YtYTfdODe","kickstart_repository_id":null,"subnet6_id":null,"pxe_loader":null,"openscap_proxy_id":null}
2016-12-16 09:18:55 - nailgun.client - DEBUG - Making HTTP DELETE request to https://foo.bar/api/v2/environments/5 with options {'verify': False, 'auth': ('baz', '*****'), 'headers': {'content-type': 'application/json'}} and no data.
2016-12-16 09:18:56 - nailgun.client - DEBUG - Received HTTP 200 response: {"id":5,"name":"9pmV69","created_at":"2016-12-16T11:18:44.522Z","updated_at":"2016-12-16T11:18:44.522Z","katello_id":null}


========================== 1 passed in 16.68 seconds ===========================
```

It is failing on sat 6.2.5 once fix of https://bugzilla.redhat.com/show_bug.cgi?id=1369737 is not cherry picked

oshtaier commented on tests/foreman/api/test_template_combination.py:
```diff
@@ -0,0 +1,80 @@
+"""
+from nailgun import entities
+
+from robottelo.decorators import tier1, skip_if_bug_open
```
put alphabetically

oshtaier commented on tests/foreman/api/test_template_combination.py:
```diff
@@ -0,0 +1,80 @@
+from robottelo.test import APITestCase
+
+
+def _delete_entities(*entities):
```
seems as pretty basic method. Maybe we can make it as a helper?

renzon replied:
Is there a place for API helpers?

oshtaier replied:
probably `utils.py` is a right place

oshtaier commented on tests/foreman/api/test_template_combination.py:
```diff
@@ -0,0 +1,80 @@
+        ent.delete()
+
+
+class TestTemplateCombination(APITestCase):
```
`class TemplateCombinationTestCase(APITestCase)`

oshtaier commented on tests/foreman/api/test_template_combination.py:
```diff
@@ -0,0 +1,80 @@
+
+
+class TestTemplateCombination(APITestCase):
+    """TemplateCombination tests"""
```
"""Implements Template Combination tests"""

renzon replied:
Fixed

oshtaier commented on tests/foreman/api/test_template_combination.py:
```diff
@@ -0,0 +1,80 @@
+                'environment_id': self.env.id
+            }])
+        self.template = self.template.create()
+        comb_dct = self.template.template_combinations[0]
```
non meaningful name for variable

oshtaier commented on tests/foreman/api/test_template_combination.py:
```diff
@@ -0,0 +1,80 @@
+            'id': comb_dct['id'],
+            'hostgroup': self.hostgroup.id
+        }
+        self.template_combination = entities.TemplateCombination(**params)
```
pretty rare usage for variable initialization, but probably it is fine

renzon replied:
I only saw it after you comment ;). Fixed as well

oshtaier commented on tests/foreman/api/test_template_combination.py:
```diff
@@ -0,0 +1,80 @@
+        _delete_entities(self.template_combination, self.template)
+
+    @tier1
+    @skip_if_bug_open('bugzilla', 1369737)
```
it is already verified, so please remove the decorator

renzon replied:
I can't remove because this is not working on sat < 6.3, so it would generate false positives there. Check: #4063

oshtaier commented on tests/foreman/api/test_template_combination.py:
```diff
@@ -0,0 +1,80 @@
+        super(TestTemplateCombination, cls).tearDownClass()
+        _delete_entities(cls.hostgroup, cls.env)
+
+    def setUp(self):
```
reviewed few times and to be honest don't see any sense in current approach. I propose to transform that setup in one good enough test case that test create and read procedure + one more that test create and delete procedure

renzon replied:
setUp is doing what it must to: creating the all entities required to test. In case someone add more tests here, no need to copy and paste entities creation logic. Plus, all entities are cleaned up after test is finished. Delete procedure itself is already tested on nailgun. Just added read test to check returned fields.

oshtaier replied:
still not sure to put all that data in setup as it will fail all tests in case something goes wrong.

oshtaier commented:
@abalakh please, review and provide your opinion. In general, I am not okay with overall structure, but need more people to check that out

abalakh commented on tests/foreman/api/test_template_combination.py:
```diff
@@ -0,0 +1,79 @@
+
+    @classmethod
+    def setUpClass(cls):
+        super(TemplateCombinationTestCase, cls).setUpClass()
```
Please add a docstring for `setUpClass()`, `tearDownClass()`, `setUp()` and `tearDown()` methods.

abalakh commented on tests/foreman/api/test_template_combination.py:
```diff
@@ -0,0 +1,79 @@
+        self.template = entities.ConfigTemplate(
+            snippet=False,
+            template_combinations=[{
+                'hostgroup_id':
```
nitpick: redundant line break here?

abalakh commented on tests/foreman/api/test_template_combination.py:
```diff
@@ -0,0 +1,79 @@
+    @skip_if_bug_open('bugzilla', 1369737)
+    def test_positive_get_combination(self):
+        """Assert API template combination get method works. Delete is
+        indirectly tested on tearDown
```
Delete is not completely tested as we're just sending the delete request and not asserting it was actually deleted (not trying to read it), suggesting a test for that case too.

renzon replied:
I am convinced with this comment. Going to add test for deletion and remove combination deletion from teardown.

renzon commented:
@abalakh @oshtaier Fixed comments. Test results:

```console
/home/renzo/PycharmProjects/venvs/robottelo/bin/python /home/renzo/bin/pycharm-2016.3/helpers/pycharm/pytestrunner.py -p pytest_teamcity /home/renzo/PycharmProjects/robottelo/tests/foreman/api/test_template_combination.py "-k TemplateCombinationTestCase"
================= 2 passed, 1 pytest-warnings in 37.67 seconds =================
```

abalakh commented on tests/foreman/api/test_template_combination.py:
```diff
@@ -77,3 +78,19 @@ def test_positive_get_combination(self):
         self.assertEqual(self.template.id, combination.config_template.id)
         self.assertEqual(self.env.id, combination.environment.id)
         self.assertEqual(self.hostgroup.id, combination.hostgroup.id)
+        combination.delete()
```
Is combination deletion really mandatory (i mean for example not deleted combination will affect other entities or lead to serious performance degradation or affect another test somehow)?
This ways tests are not isolated and in case delete functionality is broken, test for `read()` will fail too, so in case there's no real need to delete combination here other that clean things up i'd better remove this line.

renzon replied:
I had to revert it back. On sat 6.3 the get test was failing because they added a consistency constrante to not allow template removal with pending combination.

abalakh commented:
@renzon btw what does that `1 pytest-warnings` mean? Never seen message like this before.

renzon commented:
@abalakh It seem new version of Pytest is deprecating old using of fixtures:

```console
WC1 None pytest_funcarg__cov: declaring fixtures using "pytest_funcarg__" prefix is deprecated and scheduled to be removed in pytest 4.0.  Please remove the prefix and use the @pytest.fixture decorator instead.
============================== 1 tests deselected ==============================
========== 1 failed, 1 deselected, 1 pytest-warnings in 25.79 seconds ==========
```

I fixed the deleting combination on get test and added HTTPError check. ;)

abalakh commented:
@renzon thanks. you've gotten my ACK already, so all i can do is give you extra karma for that 😃

rochacbruno commented:
@oshtaier  this one is only waiting for you approval

oshtaier commented on tests/foreman/api/test_template_combination.py:
```diff
@@ -0,0 +1,97 @@
+"""
+from nailgun import entities
+from requests.exceptions import HTTPError
+from robottelo.decorators import tier1, skip_if_bug_open
```
put alphabetically

renzon replied:
Dont get this one: nailgun < requests < ribottelo

oshtaier replied:
not sure why do you look on that line, I highlighted `from robottelo.decorators import tier1, skip_if_bug_open`

oshtaier commented:
There are some changes that I treat as a basic ones, so please re-run tests and provide results

oshtaier commented:
Feel free to merge after results provided

renzon commented:
Test results:

```console
============================= test session starts ==============================
================= 2 passed, 1 pytest-warnings in 30.63 seconds =================
```

renzon commented:
Have just moved combination deletion to tearDown. Test results:

```console
================= 2 passed, 1 pytest-warnings in 25.44 seconds =================
```

oshtaier commented:
APT
